File: demo/assets/ydraw/python/yplot/smoothstep.py
#!/usr/bin/env python3
"""Conceptual reproduction of demo/scripts/yplot/smoothstep.sh — NOT RUNNABLE YET."""
from yetty.ydraw import DrawableList, Function, Plot


def show(plot):
    dlist = DrawableList()
    dlist.add(plot)
    dlist.dcs_emit()
    dlist.destroy()


# (1) Ease in/out versus a linear ramp.
print('(1) smoothstep ease vs linear ramp')
show(Plot(width=600, height=240, x_range=(0, 1), y_range=(-0.1, 1.1),
          functions=[
              Function("x", name="linear", color="#556162"),
              Function("smoothstep(0,1,x)", name="eased", color="#74C5A5"),
          ]))

# (2) Iterated smoothstep — the "smootherstep" trick.
print('(2) iterated smoothstep')
show(Plot(width=600, height=240, x_range=(0, 1), y_range=(-0.1, 1.1),
          functions=[
              Function("smoothstep(0,1,x)", name="once", color="#5A8979"),
              Function("smoothstep(0,1,smoothstep(0,1,x))", name="twice",
                       color="#74C5A5"),
          ]))

# (3) A soft window: ramp up across one edge, back down across another.
print('(3) soft window')
show(Plot(width=600, height=240, x_range=(-4, 4), y_range=(-0.1, 1.1),
          functions=[Function("smoothstep(-2,-1,x)*(1 - smoothstep(1,2,x))",
                              name="gate", color="#6BA892")]))

# (4) A soft staircase vs the hard floor() staircase.
print('(4) soft staircase vs hard floor')
show(Plot(width=600, height=240, x_range=(0, 5), y_range=(-0.2, 5.2),
          functions=[
              Function("floor(x)", name="hard", color="#364A47"),
              Function("floor(x)+smoothstep(0,1,fract(x))", name="soft",
                       color="#74C5A5"),
          ]))

# (5) A contrast curve: smoothstep between two interior edges remaps a
# 0..1 signal — the tone response of a contrast slider.
print('(5) contrast remap')
show(Plot(width=600, height=240, x_range=(0, 1), y_range=(-0.1, 1.1),
          functions=[
              Function("x", name="identity", color="#556162"),
              Function("smoothstep(0.3,0.7,x)", name="contrast",
                       color="#FFE66D"),
          ]))

File: demo/ffi/ydraw/go/yplot/smoothstep.go
// Conceptual reproduction of demo/scripts/yplot/smoothstep.sh
// NOT RUNNABLE YET — conceptual, translated from the python sketch.
package main

import (
	"fmt"

	"github.com/zokrezyl/yetty/bindings/go/ydraw"
)

func show(plot ydraw.Plot) {
	dlist := ydraw.NewDrawableList()
	dlist.Add(plot)
	dlist.DcsEmit()
	dlist.Destroy()
}

func main() {

	// (1) Ease in/out versus a linear ramp.
	fmt.Println("(1) smoothstep ease vs linear ramp")
	show(ydraw.Plot{Width: 600, Height: 240, XRange: []float64{0, 1}, YRange: []float64{-0.1, 1.1}, Functions: []ydraw.Function{
		{Body: "x", Name: "linear", Color: "#556162"},
		{Body: "smoothstep(0,1,x)", Name: "eased", Color: "#74C5A5"},
	}})

	// (2) Iterated smoothstep — the "smootherstep" trick.
	fmt.Println("(2) iterated smoothstep")
	show(ydraw.Plot{Width: 600, Height: 240, XRange: []float64{0, 1}, YRange: []float64{-0.1, 1.1}, Functions: []ydraw.Function{
		{Body: "smoothstep(0,1,x)", Name: "once", Color: "#5A8979"},
		{Body: "smoothstep(0,1,smoothstep(0,1,x))", Name: "twice", Color: "#74C5A5"},
	}})

	// (3) A soft window: ramp up across one edge, back down across another.
	fmt.Println("(3) soft window")
	show(ydraw.Plot{Width: 600, Height: 240, XRange: []float64{-4, 4}, YRange: []float64{-0.1, 1.1}, Functions: []ydraw.Function{
		{Body: "smoothstep(-2,-1,x)*(1 - smoothstep(1,2,x))", Name: "gate", Color: "#6BA892"},
	}})

	// (4) A soft staircase vs the hard floor() staircase.
	fmt.Println("(4) soft staircase vs hard floor")
	show(ydraw.Plot{Width: 600, Height: 240, XRange: []float64{0, 5}, YRange: []float64{-0.2, 5.2}, Functions: []ydraw.Function{
		{Body: "floor(x)", Name: "hard", Color: "#364A47"},
		{Body: "floor(x)+smoothstep(0,1,fract(x))", Name: "soft", Color: "#74C5A5"},
	}})

	// (5) A contrast curve: smoothstep between two interior edges remaps a
	// 0..1 signal — the tone response of a contrast slider.
	fmt.Println("(5) contrast remap")
	show(ydraw.Plot{Width: 600, Height: 240, XRange: []float64{0, 1}, YRange: []float64{-0.1, 1.1}, Functions: []ydraw.Function{
		{Body: "x", Name: "identity", Color: "#556162"},
		{Body: "smoothstep(0.3,0.7,x)", Name: "contrast", Color: "#FFE66D"},
	}})
}
